Install latest v2ray and add filtering of Tor nodes based on Speed

reece394 opened the pull request:
This PR contains some general cleanup of the downloader dockerfile and includes a script to install the latest version of v2ray based off [here](https://github.com/v2fly/alpinelinux-install-v2ray/) as the version shipped by Alpine 3.18 is 5.3 instead of the latest 15.12.1 (With Alpine 3.19 this is no longer true)

Additionally it also uses [creatorrc](https://github.com/hephaest0s/creatorrc) to help speed up tor downloads by selecting the fastest 1000 relays as guards, the fastest 1000 exists as exit, and excludes the slowest 4000 relays so they don't become middle relays. 

~~PIP install needs a separated RUN entry or else it does not work due to needing to open a new terminal window. - Needed to prevent  "The folder you are executing pip from can no longer be found."~~

reece394 committed:
Install latest v2ray and add filtering of tor nodes based on speed

sn0b4ll commented:
Hey, sorry, I don't know why Github doesn't notify me.. I will look into it in the course of the week.

As always - many thanks for the additions!

sn0b4ll commented:
Hey @reece394, sadly I can't get your fork to work, I can't connect from the controller to the download container in AriaNG.

Have you experienced such issues? Sadly from a first look I can't pinpoint from a first look, where the rpc connection breaks :/

reece394 commented:
Hmmm that is really odd. I have tested it on two systems. A 64-bit windows 10 machine and an ARM64 macOS machine both appear to be stable. 

Have you tried completely deleting the old containers from Docker and started from scratch as Docker does like to cache older versions. 

When do these issues occur so we can try different things to reproduce it?

If it is during the download process maybe with the number of connections coming in and out of the container it is getting overwhelmed???

reece394 committed:
Fix for Alpine 3.19

reece394 commented:
Updated this pull request for Alpine 3.19 support as the version of Python was upgraded. Added benefit is there is no need for the pip step anymore as stem is an apk package

sn0b4ll commented:
Hey @reece394,

I took my time to do some more testing - sadly the issues still exists for me. The problem is that the connection between the controller and downloader is failing (after setting the RPC Secret).

Steps:

1. Stop all running containers
2. RM all running containers
3. pruned all images
4. pulled your repo
5. docker compose up -d
6. Creation seems fine, both containers start up
7. go to the controller website -> can't connect to downloader

![image](https://github.com/sn0b4ll/aria2-onion-downloader/assets/1722036/5e0391bf-d346-4d7e-8442-aa9fbe99d5ae)


8. Repeated 1-3
9. Pulled current main here
10. docker compose up -d
11. go to the controller website -> instantly connects


After some further testing I saw that aria2c was never started in the download container and after waiting a couple of minutes, I get error messages from `/home/creatorrc/creatorrc.py`:
![image](https://github.com/sn0b4ll/aria2-onion-downloader/assets/1722036/9ece636c-489a-4a89-90e8-b952c7c41a88)

Seems like createtorrc tries to connect to a IP not responding anymore, blocking the creation of the config, blocking the starting of tor and therefore the aria2 downloader :smiling_face_with_tear: 

Did you already have experience with createtorrc? For me it's sadly the first time, so I don't know if there is an an easy solution to that.

sn0b4ll commented:
But there seem to be people having the same problem:
https://github.com/hephaest0s/creatorrc/issues/7

reece394 commented:
:( That actually sucks that error is happening. I have never experienced this issue. I have been using that project for about a year now and never had that problem. 

I wonder if your ISP is blocking the connection to that server. 

If you have not tried it yet try routing it through a VPN and see if that makes a difference. Will have to try and see what is happening but seems related to the stem project itself.

reece394 committed:
Fallback to fix aria-ng web ui not loading when creatorrc fails to download Descriptors

reece394 commented:
I have not fixed the download issue however I added a fallback to the script when it does not work to use the current working tor command so it should work the same way as before if the download is not successful. This is probably the best solution until we can figure out why they are being blocked on your end.

pocoloko commented:
FWIW I tried the updated repo from @reece394 and it works fine

docker-compose started fine, the log indicated creatorrc was successful, connected to controller and it connected fine to the downloader, tried a download and it started and worked fine.

Docker version 24.0.7
Docker Compose version 2.23.3

sn0b4ll commented:
Sorry, I have a lot of private stuff going on right now, sorry for keeping you waiting. 

Thanks for implemented the fallback @reece394 and also thanks @pocoloko for testing - I will test it myself this weekend and see if the fallback works in my region.

sn0b4ll commented:
Works great now, many thanks!
